[4.0] Add support for glow maps

Ansraer opened the pull request:
- *Production edit: 4.x version of https://github.com/godotengine/godot/pull/93133.*

This PR is a minor enhancement for the glow post processing effect and allows users to provide a glow map (e.g. lens dust).
The influence of this glow map can be controlled with the glow_map_strength float variable, which defaults to 0 (=glow map turned off).

Here is an example screenshot I took while using a low res lens-dust jpg I found on google as the glow map:
![Screenshot from 2021-11-04 00-01-55](https://user-images.githubusercontent.com/6329420/140231734-248e5625-f42d-4eed-8e9b-9ed523bdde28.png)

~~**Disclaimer**: I originally wrote this for one of my projects, before deciding to turn it into a PR. I did my best to clean things up as much as I could, but in the processes ended up nearly doubling the number of tone mapping pipelines. Would greatly appreciate some feedback on that part in particular from someone with more rendering experience.~~
Nevermind, after discussing it on rocketchat that change was reverted. The new `#ifdef` blocks were removed and the glow map was moved to binding 1 of set 2.

Calinou commented:
> The influence of this glow map can be controlled with the glow_map_strength float variable, which defaults to 0 (=glow map turned off).

Shouldn't it default to `1` or perhaps `0.5`, with the effect being automatically nullified if no glow map texture is specified? This way, you can see an effect immediately as soon as a glow map texture is loaded.

Ansraer commented:
As discussed with @clayjohn on rocketchat I removed the `#define` and moved the glow map into set 2 binding 1.

> > The influence of this glow map can be controlled with the glow_map_strength float variable, which defaults to 0 (=glow map turned off).
> 
> Shouldn't it default to `1` or perhaps `0.5`, with the effect being automatically nullified if no glow map texture is specified? This way, you can see an effect immediately as soon as a glow map texture is loaded.

The effect already turns itself of if no glow map is provided, but I can also change the default value. I think I will set it to 0.8 instead of 1, that is what I have been using during testing so far.

Ansraer commented:
Huh, must have lost the docs during the rebase. Should be fixed now.

DarkKilauea commented on doc/classes/Environment.xml:
```diff
@@ -154,6 +154,12 @@
 			The intensity of the 7th level of glow. This is the most "global" level (blurriest).
 		</member>
+		<member name="glow_map" type="Texture" setter="set_glow_map" getter="get_glow_map">
+			The texture that should be used to as glow map.
```
```suggestion
			The texture that should be used as a glow map.
```

Ansraer replied:
fixed

Calinou commented:
I tested this feature locally, it's working pretty well overall:

![image](https://user-images.githubusercontent.com/180032/140626617-61697dc1-9a74-4b14-a1e9-49767a5ca31f.png)

**Testing project:** [test_glow_map.zip](https://github.com/godotengine/godot/files/7491395/test_glow_map.zip)

Two things I noticed:

- When the glow **Mix** blend mode is used, the glow map is ignored. Ideally, it should be possible to use the glow map while the Mix mode is active.
- The texture is stretched to fit the screen. This is usually fine, but the stretching may look unpleasant at ultrawide aspect ratios. This is especially the case if you're using a texture with a 1:1 aspect ratio as a glowmap (which I did in my testing project). There are two ways to alleviate this:
  - Recommend using a glow map with a 16:9 aspect ratio. This covers the most common aspect ratio without stretching, but it will still appear to stretch noticeably on 21:9 monitors.
  - Only use the center of the glow map in a way similar to the CSS `background-size: cover` property. This means stretching will no longer occur on aspect ratios wider than the glow map, but the glow map will appear zoomed in instead. I'm not sure if this is actually a better choice, but it's something to keep in mind.

Calinou commented on servers/rendering/renderer_rd/shaders/tonemap.glsl:
```diff
@@ -425,9 +426,10 @@ void main() {
-
 	if (params.use_glow && params.glow_mode != GLOW_MODE_MIX) {
 		vec3 glow = gather_glow(source_glow, uv_interp) * params.glow_intensity * params.luminance_multiplier;
+		if (params.glow_map_strength > 0.001)
+			glow = mix(glow, texture(glow_map, uv_interp).rgb * glow, params.glow_map_strength);
```
As per the code style guidelines, braces are now required for single-line `if` statements:

```suggestion
		if (params.glow_map_strength > 0.001) {
			glow = mix(glow, texture(glow_map, uv_interp).rgb * glow, params.glow_map_strength);
		}
```

Ansraer replied:
I would have expected clang format to fix that. Good to know.

Ansraer commented:
Fixed the doc typo, mix mode (that was a stupid oversight, no clue how I missed that) and added braces to the if statements.

> The texture is stretched to fit the screen. This is usually fine, but the stretching may look unpleasant at ultrawide aspect ratios. This is especially the case if you're using a texture with a 1:1 aspect ratio as a glowmap

In theory I could add another variable to the Params uniform to adjust the UV based on the texture-vs-screen ratio. However, I am not sure if that is an ideal solution. There is other stuff that will change on monitors with unusual resolution and aspect ratios as well, so it might be better if developers write their own logic to swap out the glow map depending on the resolution.
I am fairly certain I already exposed the necessary function to scripting for just that reason.

Calinou commented on doc/classes/Environment.xml:
```diff
@@ -154,6 +154,12 @@
+			The texture that should be used as a glow map.
+		</member>
+		<member name="glow_map_strength" type="float" setter="set_glow_map_strength" getter="get_glow_map_strength" default="0.8">
+			How strong of an impact the glow map should have on the overall glow effect.
```
```suggestion
			How strong of an impact the [member glow_map] should have on the overall glow effect. A strength of [code]0.0[/code] means the glow map has no effect on the overall glow effect. A strength of [code]1.0[/code] means the glow has a full effect on the overall glow effect (and can turn off glow entirely in specific areas of the screen if the glow map has black areas).
```

Calinou commented on doc/classes/Environment.xml:
```diff
@@ -154,6 +154,12 @@
 			The intensity of the 7th level of glow. This is the most "global" level (blurriest).
 		</member>
+		<member name="glow_map" type="Texture" setter="set_glow_map" getter="get_glow_map">
+			The texture that should be used as a glow map.
```
```suggestion
			The texture that should be used as a glow map to [i]multiply[/i] the resulting glow color according to [member glow_map_strength]. The texture's RGB color channels are used for modulation, but the alpha channel is ignored.
			[b]Note:[/b] The texture will be stretched to fit the screen. Therefore, it's recommended to use a texture with an aspect ratio that matches your project's base aspect ratio (typically 16:9).
```

Ansraer commented:
Updated the PR with the suggested documentation changes. And yeah, colored glow maps is something I really want to explore myself in the future. I suspect that using three different voronoi id (not distance) noise textures in the rgb channels of a glow map would make for a really trippy effect. Maybe even make those textures tiling and somehow animate them?

Ansraer committed:
add support for glow maps

akien-mga commented:
Thanks!
